HEAD response bodies are stripped before fairings are run

teuron opened the pull request:
fixed #546 

What needs to be implemented

- [x]  Test: Fairings should only be called once per request/response pair.
- [x]  Test: Response fairings see a `GET` request for auto-handled `HEAD` requests.
- [x]  Request fairings should be called with the original request.
- [x]  Update the docs to clarify the new constraints.

jebrosen commented:
This will still strip response bodies for autohandled routes. I think that aspect could be fixed by changing line 230 from `request.strip_body()`  to `request.set_method(Method::Head);

teuron commented:
Thanks, but Rust thinks here that request is still borrowed. I choose this solution but I could also make an unsafe block here and create a new reference. What do you think is better?

teuron commented:
And is it actually wanted, that the on_response fairings are called twice on autohandeled HEAD request?

jebrosen commented:
Without seeing the error message, I'm guessing that request is actually borrowed by the response. A perhaps cleaner solution would be to stash something like `let was_head_request = (request.method() == Method::Head);` early on and then check against `was_head_request` to strip the body or not -- that reduces the special-casing of auto-handled HEAD.

That would help with strip_body, but there is another issue I just noticed: when a HEAD request is auto-handled, response fairings are told the method was GET. IMO This behavior should either be an explicit design choice or a bug that gets fixed.

jebrosen commented:
I posted the previous before seeing your other comment about fairings being called twice. The recursion for auto-handled HEAD certainly makes everything harder to reason about... now I'm unsure of the right way to address all this.

If I'm reading right, the following happens for auto-handled requests:
1. GET route runs
2. fairings run against the GET request
3. body is stripped
4. fairings run again, this time against the stripped GET request <-- oops
5. response is returned
In this case, (4.) should not happen.

And for "real" head requests:
1. HEAD route runs
2. body is stripped
3. fairings run against the stripped HEAD request
4. response is returned
In this case, (2) and (3) should probably be switched, if only so that fairings cannot add bodies to head request.

teuron commented:
My proposed solution would be this workflow: 

Auto-handeled HEAD Requests: 

1. Head request fairings run
2. GET route runs without any fairings
3. Head response fairings run
4. Response body is stripped
5. Return response

Normal HEAD Requests: 

1. HEAD route runs
2. fairings run against HEAD request
3. Response body is stripped
4. Return response

teuron commented:
@jebrosen I implemented a working solution with my proposed workflow.

SergioBenitez commented:
Looks like we've identified issues with the existing implementation: awesome!

Here's what _should_ happen:
  * Fairings should be called _once_ for a given request/response pair.
  * Fairings should be called with the original request (post-[`preprocess`](https://github.com/SergioBenitez/Rocket/blob/master/lib/src/rocket.rs#L172)).
    - This means requests fairings should see a `HEAD` regardless of what Rocket does thereafter.
    - Bodies should be stripped on incoming `HEAD` requests  _even if_ a fairing changes the method.
  * Response fairings should see the `Request` the passed in `Response` was generated for.
    - This is consistent with response fairings seeing changes made by other request fairings.
    - This means response fairings see a `GET` request for auto-handled `HEAD` requests.

This PR and #546 add the following constraint:
  * Fairings should be called with the final response _before_ body stripping.

I am in favor of this change, and I'm certainly in favor of fixing the existing implementation so that it always does what should have been happening. @teuron I feel that your current implementation is a bit hard to follow. We need to ensure that the `dispatch` method is easy to understand.

Note that this PR will also need to fix the documentation on fairings and automatic head handling where appropriate to clarify these semantics.

SergioBenitez commented:
@jebrosen Thank you for the commentary on this PR. It's very helpful and appreciated!

jebrosen commented:
@SergioBenitez that analysis/solution sounds good to me, and I also agree that this implementation is a bit hard to follow as-is. The latest version also adds a `transmute` that appears to create two `&mut` references to the same request, which I am very uncomfortable with. Hopefully there is a simpler way to achieve all the goals outlined above.

I will mention that I have changed my own code so that this and #546 are no longer blocking me personally, but what I did only works on autohandled requests and only for fairings that can tell from the response they have been run once already.

teuron commented:
@SergioBenitez @jebrosen I've implemented everything based on your suggestion. 

I've added some notes in the docs, and I think that would be enough. 


Thanks for your support!

SergioBenitez commented:
@jebrosen Would you be able to review the updated code?

jebrosen commented:
I have read the latest code and it "looks right" to me, for whatever that's worth. I would like to test it more thoroughly, but unfortunately will not have the time for a while.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -191,17 +191,33 @@ impl Rocket {
+        let was_head_request = request.method() == Method::Head;
 
         // Route the request to get a response.
         let mut response = match self.route(request, data) {
```
I think the better refactor is to split this big `match` out into another method, put everything before and after it in `dispatch`, and call the new method in the autohandled case instead of dispatch. In psuedocode:

```rust
fn dispatch(request, data) -> Response<'r> {
    // Remember whether this was originally a HEAD request or not.
    let head_request = request.method().is_head();

    // Preprocess the request and run the request fairings.
    self.preprocess_request(request, &data);
    self.fairings.handle_request(request, &data);

    // Route the request and run the user's handlers.
    let mut response = self.route_and_process(request, &data);

    // Run response fairings and strip the body if this was a HEAD request.
    self.fairings.handle_response(request, &mut response);
    if head_request { response.strip_body(); }

    response
}

fn route_and_process(request, data) -> Response<'r> {
    match self.route(request, data) {
        ...
        Outcome::Forward(data) => {
            if request.method() == Method::Head {
                request.set_method(Method::Get);
                self.route_and_process(request, data)
            } else {
                self.handle_error(Status::NotFound, request)
            }
        }
        ...
    }
}
```

This removes all of this `if` statements, clarifies the logic, and separates concerns nicely.

SergioBenitez replied:
Both this PR and the existing implementation were calling `self.preprocess_request()` twice as well. The refactor I propose makes it clear to see that this doesn't happen.

teuron replied:
@SergioBenitez thanks thats the best way we can do it. I will implement it next week and update the PR.

jebrosen commented on lib/src/rocket.rs:
```diff
@@ -191,20 +191,49 @@ impl Rocket {
+
+    fn route_and_process<'s, 'r>(
+        &'s self,
+        request: &'r Request<'s>,
```
This should be `&'r mut` instead of just `&'r`, since this function might modify `request`

SergioBenitez replied:
Absolutely! Great catch! This mistake leads to highly undefined behavior when the `&mut Request` is created later from it.

jebrosen commented on lib/src/rocket.rs:
```diff
@@ -214,20 +243,24 @@ impl Rocket {
+                    // Rust thinks `request` is still borrowed here, but it's
+                    // obviously not (data has nothing to do with it), so we
+                    // convince it to give us another mutable reference.
+                    // TODO: Use something that is well defined, like UnsafeCell.
```
What is the status of this TODO -- is there a better solution than the `unsafe` block? NLL is available now.

SergioBenitez replied:
I'm not sure if NLL resolves this particular issue, but I'd be curious to know if it does. That being said, there are far too many soundness problems with the NLL implementation as it stands for me to be comfortable enabling it in Rocket.

teuron replied:
I tried to use it with NLL, but it does not solve the problem. Do you have any suggestions?

SergioBenitez replied:
You can leave this as is.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -191,20 +191,47 @@ impl Rocket {
+        // Add the 'rocket' server header to the response and run fairings.
+        // TODO: If removing Hyper, write out `Date` header too.
+        response.set_header(Header::new("Server", "Rocket"));
+
```
Remove this empty line.

SergioBenitez commented on lib/src/request/request.rs:
```diff
@@ -616,6 +616,12 @@ impl<'r> Request<'r> {
 
+    /// Set the method of `self`, even if self is a shared reference
+    #[inline(always)]
+    pub(crate) fn set_method_interior(&self, method: Method) {
```
Let's name this `_set_method()`. We should use `_method_name()` when we have an internal-only method that conflicts with a public method.

SergioBenitez commented on lib/src/request/request.rs:
```diff
@@ -446,7 +446,7 @@ impl<'r> Request<'r> {
     pub fn format(&self) -> Option<&MediaType> {
         static ANY: MediaType = MediaType::Any;
-        if self.method.supports_payload() {
+        if self.method.get().supports_payload() {
```
Don't use `self.method.get()`. Instead, use `self.method()`. Make this change everywhere else `self.method.get()` is used.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -214,20 +241,16 @@ impl Rocket {
-                    let mut response = self.dispatch(request, data);
-                    response.strip_body();
+
+                    request.set_method_interior(Method::Get);
```
Move this line just after the next comment with no empty line before or after it.

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,74 @@
+
+use rocket::response::content;
+
+
```
Remove this empty line.

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,74 @@
+        let rocket = rocket::ignite()
+            .mount("/", routes![index])
+            .attach(AdHoc::on_response(|req, res| {
+                if req.method() != Method::Head || res.body_string() != Some("{ 'test': 'dont strip before fairing' }".into()) {
```
Can we simply `assert!` here and cause the test to fail by panicking?

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,74 @@
+                 let c = req.guard::<State<Counter>>().unwrap();
+
+                 if c.get.fetch_add(1, Ordering::Release) != 0 {
+                     req.set_method(Method::Put); //On Error set bad Method
```
Comment should be `// This is the comment.`

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,74 @@
+                 }
+            }))
+            .attach(AdHoc::on_response(|req, res| {
+                if req.method() != Method::Get || res.body_string() != Some("{ 'test': 'dont strip before fairing' }".into()) {
```
Same with the `assert!` here.

SergioBenitez commented on site/guide/fairings.md:
```diff
@@ -96,7 +96,8 @@ events is described below:
     applications fails to by rewriting **404** responses as desired. As another
     example, response fairings can also be used to inject headers into all
-    outgoing responses.
+    outgoing responses. Beware that the body of a `HEAD` request will be
```
This seems very out of context.

If you'd like, we can focus on the code for this PR and I can take care of the documentation changes afterwards.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -214,20 +240,15 @@ impl Rocket {
+
+                    // Dispatch the request again with Method `GET`.
+                    request._set_method(Method::Get);
+                    let response = self.route_and_process(request, data);
```
Remove the `let` and just return `self.route_and_process(request, data)` directly from this branch.

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,69 @@
+
+#[head("/")]
+fn index() -> content::Json<&'static str> {
+    content::Json("{ 'test': 'dont strip before fairing' }")
```
Can we move this string into a `const` so that we don't keep repeating it?

SergioBenitez replied:
Also, is there a reason you're returning `Json` instead of just a `'static str`? It's better to keep it simple and focused on what we're trying to test.

SergioBenitez commented on lib/tests/fairing_before_head_strip-issue-546.rs:
```diff
@@ -0,0 +1,69 @@
+
+        let client = Client::new(rocket).unwrap();
+        let response = client.head("/").dispatch();
+
```
Remove this empty line. Same in the test below.

SergioBenitez commented:
Merged in 3a1d6718946a0a3830ffff3a7b15faf9d48182e5. Super awesome work, @teuron!
